### ISSUE TITLE
[Enhancement] UWP as a first-class platform (alongside Android and iOS)

### ISSUE THREAD

**Issue by MuleaneEve** [dotnet/maui#158]
## Summary

I would like UWP to be developed and supported at the same level as Android and iOS.

In Xamarin.Forms, it is very common for samples to be made only for Android and iOS. And worse, it often happens that new features are launched without UWP support at first (and some times, that support stays incomplete for a long time. See Shell...).

I hope that the new [Single Project proposal](https://github.com/dotnet/maui/issues/70) will make this request more seamless, but I would also like the developers working on MAUI to give a bit more consideration to UWP.

## Changes

1. Same API support for all three platforms at the same time.

Edit: Based on [this comment](https://github.com/dotnet/maui/issues/158#issuecomment-662795172):

2. Make MAUI code as platform-independent as possible. That way, even if a developer only writes for Android/iOS. The code will implicitly work on other platforms as well. For example, reducing the reliance on renderers would be a good step in that direction.

3. Make it possible for Xamarin developers to compile & run UWP apps on macOS (with a remote connection like it is currently possible to compile iOS apps on Windows).

## Intended Use Case

If Microsoft itself neglects UWP in favor of the other platforms, it doesn't help UWP stay relevant.

**Comment by Happypig375** [comment]
And since Microsoft is supporting macOS officially in MAUI, the same should also apply to macOS.

**Comment by jcbeppler** [comment]
This looks a lot like this huge issue, where people would like MAUI to be made based/forked on UWP and not on Xamarin:
Duplicate of #https://github.com/dotnet/maui/issues/43

**Comment by saint4eva** [comment]
> This looks a lot like this huge issue, where people would like MAUI to be made based/forked on UWP and not on Xamarin:
> Duplicate of ##43

What @MuleaneEve MuleaneEve was requesting is that more effort be given to make sure MAUI has a proper support for UWP - in terms of features and tooling. 

Stop mixing XAML debates with this.

**Comment by jcbeppler** [comment]
@saint4eva I don't think is mixing the debates, he wants something towards WPF, so I think is the same thing. Everyone wants a base which is either WPF or UWP. So far it has been proposed Xamarin which is something no one likes (sorry, not counting students).

**Comment by MuleaneEve** [comment]
Within the context of this issue, I don't care what XAML/architecture MAUI uses. I just want UWP to be well supported.

Let's keep each issue focused on one thing; it will make it easier for the developers to triage and discuss them.

**Comment by MhAllan** [comment]
I sometimes think that cross platform UI with this UWP thing is not possible. here is example

```xaml
<StackLayout Padding="30">
		<Frame Padding="40, 0"
			   CornerRadius="30"
			   BorderColor="LightBlue">
			<Picker Title="Something"
					ItemsSource="{Binding Options}"
					BackgroundColor="Transparent" />
		</Frame>
	</StackLayout>
```

See result below .. I just can't understand what is that, and when I look at it I can't stop wondering, where in computer UI history we had picker with it's placeholder (or title) on top of the control.

Now if this is UWP issue, then cross platform UI with this UWP is impossible. If this is Xamarin implementation, please don't implement UWP anymore if this is the result.

As you see, for this code (frame, and picker) we need OnPlatform to fix the problem. If we need one OnPlatform switch per 2 controls, then we rather have one OnPlatform per whole page, and forget about shared UI.

![Capture](https://user-images.githubusercontent.com/13087546/85696466-51595e00-b71c-11ea-8351-f4fcb4e66790.PNG)

**Comment by JohnGalt1717** [comment]
I'd propose that instead of trying to make Xamarin Forms (ne Maui) work well with UWP, Microsoft should just dump Maui and use Flutter rendering and put the effort into getting flutter working on UWP. See here: https://github.com/dotnet/maui/issues/160

**Comment by davidortinau** [comment]
> where in computer UI history we had picker with it's placeholder (or title) on top of the control.

@MhAllan that's a native [UWP ComboBox](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/combo-box).

The `Picker` implementations in Xamarin.Forms are composite controls due to how they are used in a mobile first way. This is a good topic for discussing how we move these controls forward for desktop.

https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.UAP/PickerRenderer.cs

Proposals?

**Comment by saint4eva** [comment]
> I'd propose that instead of trying to make Xamarin Forms (ne Maui) work well with UWP, Microsoft should just dump Maui and use Flutter rendering and put the effort into getting flutter working on UWP. See here: #160

Are you being paid to write the same comments throughout this repo? Most times, out of context of the topic being discussed.

**Comment by JohnGalt1717** [comment]
1. This isn't out of context. If the flutter rendering engine is adopted and works on windows, then UWP would be part of that package and #160 clearly mentions making plugins to flutter with UWP.  And then the issues in this issue go away for the most part because of how it would be structured.

2. No I'm not paid. I'm attempting to get eyes on a serious issue with Microsoft's lack of vision, and provide the powers that be with a real solution that saves them in the end user space.  I want Microsoft to succeed. The current approach clearly won't. It hasn't in the last 10 years of Xamarin, and there is nothing about Maui that indicates that that is going to change.

**Comment by saint4eva** [comment]
> 1. This isn't out of context. If the flutter rendering engine is adopted and works on windows, then UWP would be part of that package and #160 clearly mentions making plugins to flutter with UWP.  And then the issues in this issue go away for the most part because of how it would be structured.
> 2. No I'm not paid. I'm attempting to get eyes on a serious issue with Microsoft's lack of vision, and provide the powers that be with a real solution that saves them in the end user space.  I want Microsoft to succeed. The current approach clearly won't. It hasn't in the last 10 years of Xamarin, and there is nothing about Maui that indicates that that is going to change.

I believe I have been vindicated.

**Comment by jcbeppler** [comment]
@JohnGalt1717 OK, just you have vision and everyone else is wrong. 😄  
I'm pretty sure you can go to Flutter repository and ask them to create something for Windows there.

**Comment by MhAllan** [comment]
@davidortinau 
> Proposals?

Before proposals, It is not UWP issue, it is actually Xamarin implementation. 

Look here: https://github.com/xamarin/Xamarin.Forms/issues/3646 

Does the Xamarin guy sound right? for me he is 100% wrong. A placeholder is were a placeholder is, and it is not meant to be visible when there is data. Your decision in Xamarin is what actually made this control look strange.

**Comment by jcbeppler** [comment]
@MhAllan I must say, this MAUI made my life much funnier... now I laugh everyday because I always thought I was the only one who didn't like Xamarin... :joy: :joy: :joy:

**Comment by bmacombe** [comment]
@MhAllan I'm confused why you think XF 3646 is the wrong way.  It seems to me Hartez makes sense.  XF Title prop to UWP ComboBox.Header makes sense than placeholder as a default.  Using place holders is a style choice that not everyone wants.  

If the XF Picker property was "Placeholder" and was being mapped to the UWP "Title" property I would understand the confusion.

The great thing with forms is it doesn't take much to create a custom renderer that would change that behavior if you like, as mentioned in that PR.  With MAUI making those custom renderers will be even easier it sounds like.

**Comment by MitchBomcanhao** [comment]
> If Microsoft itself neglects UWP in favor of the other platforms, it doesn't help UWP stay relevant.

Good UWP support would help Xamarin/MAUI stay relevant compared to other multiplatform development systems that completely ignore UWP.
Pushing back UWP development because "not many people use it" is pretty much why people don't use it in xamarin.
Think about your unique selling points - here's a clue: doing the same android+ios apps as other systems isn't it.

We started life with native windows phone and android apps.
This extended to a native windows+windowsPhone app, and Xamarin.Forms android+iOS.
Windows support was never good enough on Xamarin forms, and with the death of WinRT apps and UWP being the future, we thought that using UWP via Xamarin forms was the only way to keep a windows app in the mix - it would definitely not be made native again.
Over time UWP has gotten better, but Xamarin have started replacing the existing stuff with the new shell paradigm. UWP was left well behind in Shell support. People who want to support UWP can't pick up the new toys because they either don't exist or don't work as expected, with experimental and buggy stuff all around.
Even simpler new features often don't come to UWP until much later (if at all) or are broken.

People who create demo apps and do videos on these new features often completely ignore UWP variants. sometimes those variants exist in the solutions, but the app will crash when you run it. Or, also often, the feature that is being demoed doesn't even work at all. This just shows how little attention is given to UWP.

Even basic stuff like tooling is constantly broken because VS for windows and VS for mac often have different versions of xamarin related components, which means that connecting a windows machine to a mac is very unreliable - updating visual studio across platforms is indeed a lottery. I recall this new xamarin related feature for VS2019 that got blog posts and everything - "go do this new thing, it is great". Yeah I tried it and it absolutely didn't work. further digging into the obscure Microsoft support pages it became apparent that the new feature only worked on macs because of the way different systems deal with files. Nobody at Microsoft tried it on windows? dear god.

I wonder how many people at microsoft just run their stuff on their macs, so don't ever have any issues with VS2019 being inconsistent across platforms, and also don't bother with UWP because they can't even make or build UWP projects from the mac (correct me if I'm wrong).

Sorry for the rant.

**Comment by freever** [comment]
@MitchBomcanhao I totally agree. The Microsoft approach to UWP is my number one frustration with Xamarin.Forms as a platform. As you point out, the XF team has been banging on about how wonderful Shell is, but it is unusable on UWP. That is not even hyperbole - navigation as a concept has just not been implemented in the UWP version of Shell - https://github.com/xamarin/Xamarin.Forms/issues/8498. It isn't a bug, it's how it has been implemented. 

I care about UWP 1) because it's the quickest development loop on a windows machine and 2) I want to ship admin functionality as a windows app so I don't have to build a web app, and that means more code reuse for me. Long story short, I don't use shell, and will not until it works on UWP.

**Comment by jcbeppler** [comment]
@MitchBomcanhao @freever Long live to the native UWP! (Not Xamarin) 😂

**Comment by MitchBomcanhao** [comment]
Just remembered one more thing - Hot reload (or whatever its called).

Xamarin teams have put in a lot of work to get it working and updating almost in real time for iOS and Android. No such luck for Xamarin UWP. Interestingly, native windows store apps have had hot reload for years.
Is this missing from Xamarin UWP because:
- there is some sort of technical issue that prevents it - its a feature that already exists in native UWP so one would think this wouldn't be too complicated to get up and running on Xamarin forms apps.
- somebody thought of getting it done but their boss told them "if its not iOS or Android we don't care".

People talk about rapid prototyping and development, what's quicker than developing on the platform that doesn't require android virtual machines, or devices attached to the computer, or connecting to random macs and running those horrible iOS simulators... So much easier to do it all in windows, but the people who make windows and the development tools keep forcing us NOT to use it.

**Comment by MhAllan** [comment]
@bmacombe Thanks for clarifying the difference between Title and Placeholder, that makes sense now. so we can say, I am using the "Title" thinking it is same to "Placeholder" but Xamarin forms doesn't have Placeholder for the Picker.

I also now understand @davidortinau saying "how we move these controls forward for desktop"

Interesting, how about the decision was that we implement a Placeholder that maps to native Placeholder and if there is no native placeholder then it will map to the native Title. wouldn't that lead to a more desired result?

MAUI should have way to make things have similar look on different platforms, Visual = "Material" could be implemented for UWP maybe?

**Comment by dersia** [comment]
I have been always advocating for having uwp as a Frist-class citizen, but I think one very important feature in the tooling is missing for this.

There are contributors working on XF/MAUI features who use macs for development. 
As long as there is no tooling for those to at least build, yet better test, uwp applications from a Mac uwp will always stay a second-class citizen. 

I asked a while back for having tolling to support uwp the same way tooling is allowing ios and macos development on a windows machine through ssh connection. 

The same can be made available through rdp or even ssh on newer windows versions. And uwp has have had a simulator for years now and could also be streamed to a Mac. 

It would even be easier for Mac users, since they can just run windows in a vm on their Mac and then stream the simulator or use ssh with the build tools. Windows users will always need a accessible Mac-hardware. 

So for this issue to have a affect, we need to start at tooling!

**Comment by Happypig375** [comment]
Uno can run UWP on Macs.

**Comment by evolcano** [comment]
I hope MAUI is a new and solid start point to solve the problems target Windows and WEB, beside iOS and Android.
I hope EVERYONE in Microsoft can be united efficiently, break the boundary of product and origination,  set the common goal base on market and technology.
I hope I could use Shell in Xamarin as @freever said, and upgrade my project and library to .Net Standard 2.1 and NetCore 3.* in the nearly future. 
I hope we can survive and do a great business along this technology roadmap base on Xamarin/MAUI.

**Comment by johnshardman** [comment]
I completely agree with the sentiment that the Windows target should be treated as a first class citizen (note that it looks like it will be WinUI rather than UWP support in MAUI). However, my concern is that the roadmap shown at https://github.com/dotnet/maui/wiki/Roadmap has Mac parity scheduled before Windows parity. Despite the number of machines running Windows being greater than the number of Macs, it appears that the long-standing preference of Macs among  the former Xamarin team is still having too much influence on the rollout of support that customers actually need. The roadmap certainly makes it look like Windows support has slipped from third on the list to fourth.

**Comment by jcbeppler** [comment]
I was a little bit skeptic regarding UNO but looks like they know better than MSFT on which path to go.
By the way, why do people make threads in github? Has anyone ever saw a corporation changing a decision because of public opinion?

**Comment by MuleaneEve** [comment]
@jcbeppler I understand that is is easy to be skeptical, but the .NET team has changed their plans many times based on public feedback.
Of the top of my head, I can think of two examples:

1. The first versions of ASP.NET Core used to be able to run on top of .NET Framework. This was important during the transition period. And Microsoft had initially planned to drop that support with .NET Core 2.0. However, based on public feedback, [they moved that plan to .NET Core 3.0](https://docs.microsoft.com/en-us/dotnet/core/compatibility/2.2-3.0#target-framework-net-framework-support-dropped).
2. EF Core 3.0 dropped support for .NET Standard 2.0, which made it not compatible with .NET Framework and UWP. But [they added that support back with EF Core 3.1](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-3.0/breaking-changes#netstandard21).

So, I am not saying that Microsoft will always listen to whatever the "public" asks (which is impossible), but let's give them constructive feedback. That's the best way to steer them in the direction we want.

**Comment by bmacombe** [comment]
I suspect the MacOS platform is currently less of a first-class citizen compared to UWP at this moment and will require more work then UWP to bring iOS/Android parity.  So it would make sense to put it sooner in the development plan than UWP, during the "Preview" period.  Windows is still scheduled to be done before the initial release.  Let's not attribute to preferences and politics what engineering needs can explain.

**Comment by jcbeppler** [comment]
@MuleaneEve I mentioned I was initially skeptical about UNO Platform, not MAUI. MAUI I am pretty sure it's the most horrible announcement I've heard since I started using C# (around 12 years ago). Not because the one UI is a bad idea, it's because using Xamarin is a bad/horrible/creepy idea. (I'm lacking enough "bad" synonyms here and yeah, I'm pissed as many people are pissed with these)

Nowadays I think the UNO Platform guys are doing exactly what every C# developer expected in his whole life, XAML+C# which can be used on everything: Windows desktop (the only desktop OS people care about), Web, Android and iOS. Not this Xamarin/MAUI hybrid monster which aggregates no value.

And to be honest, in the years since I started paying programming, I've seen little improvement in the tools I use. 
Ok, .NET Core which runs on Linux ok, cool but except that, what? Oh, ASP.NET, why the hell I still need to use HTML/CSS instead of XAML? Is it something to punish the developers so they need to learn many coding standards? Is it too hard to use one thing everywhere? The UNO guys are doing that, so I would expect this as the minimum level from a corporation.

And I'm not asking MSFT to listen everything the public asks, as this is impossible because each one has one opinion. I'm asking to listen to the majority.

**Comment by KPixel** [comment]
I think that the main reason why UWP is harder to support for Xamarin developers is iOS:
Since, they must use macOS to support iOS; it requires extra efforts to also have a Windows environment for UWP.

This is also the case for third-party libraries developers like the [author of PancakeView who blogged about that](https://www.thewissen.io/pancakeview-goes-v2-0-heres-whats-new/).

So, I can think of two (complementary) solutions to this issue:

1. Make MAUI code as platform-independent as possible. That way, even if a developer only writes for Android/iOS. The code will implicitly work on other platforms as well. For example, reducing the reliance on renderers would be a good step in that direction.
2. Make it possible to compile & run UWP apps on macOS (with a remote connection like it is currently possible to compile iOS apps on Windows).

**Comment by Happypig375** [comment]
In addition to the Remoted iOS Simulator, we should have Remoted macOS app deployment. This way, we can develop on all platforms with Windows or Mac.

**Comment by LeftTwixWand** [comment]
I think UWP'll die in a few years. Project Reunion - future of desktop development.

**Comment by MuleaneEve** [comment]
@LeftTwixWand Within the context of this issue, Project Reunion doesn't change anything. At best, MAUI will switch from using UWP controls to using WinUI 3 controls.

**Comment by saint4eva** [comment]
> @MuleaneEve I mentioned I was initially skeptical about UNO Platform, not MAUI. MAUI I am pretty sure it's the most horrible announcement I've heard since I started using C# (around 12 years ago). Not because the one UI is a bad idea, it's because using Xamarin is a bad/horrible/creepy idea. (I'm lacking enough "bad" synonyms here and yeah, I'm pissed as many people are pissed with these)
> 
> Nowadays I think the UNO Platform guys are doing exactly what every C# developer expected in his whole life, XAML+C# which can be used on everything: Windows desktop (the only desktop OS people care about), Web, Android and iOS. Not this Xamarin/MAUI hybrid monster which aggregates no value.
> 
> And to be honest, in the years since I started programming, I've seen little improvement in the tools I use.
> Ok, .NET Core which runs on Linux ok, cool but except that, what? Oh, ASP.NET, why the hell I still need to use HTML/CSS instead of XAML? Is it something to punish the developers so they need to learn many coding standards? Is it too hard to use one thing everywhere? The UNO guys are doing that, so I would expect this as the minimum level from a corporation.

Then use UNO Platform, if it meets your need. Kindly note that .NET is not only for XAML developers.

**Comment by jcbeppler** [comment]
> Then use UNO Platform, if it meets your need. Kindly note that .NET is not only for XAML developers.

@saint4eva I mentioned that the UNO Platform is the bare minimum I expect. And I mean minimum, so I am not sure if my expectations are too high or the MSFT managers expectations are too low.

And I agree, is not only for XAML developers but is the "expected idea" that with one programming language you can create software for multiple targets... at least reuse parts of the software...  be it with xaml or html/css. XAML looks a little bit more organized to my eyes but I also like some things from html/css. 
Having both is like saying to a web developer: For web you use C#/html/css but for desktop you will have to use Assembly/html/css because you know, that's it. It doesn't make sense.,

It must have a standard combination and go into one direction.

**Comment by trevelyanuk** [comment]
I suppose rather than say "**UWP** as a first-class platform", I would consider saying "**Windows** as a first-class platform".

From my own point of view, I'm more concerned that I can write a mobile app that will run on iOS/Android/Windows than really what "kind" of application that ends up being. If UWP gets ditched in favour of some Win32 application, or Project Reunion introduces some other platform technology, I am not particularly bothered - so long as I can write as little platform-specific code as possible. 

As an evolution of Xamarin Forms, MAUI should be aiming to have a platform-agnostic approach as much as possible with a focus really on having a UI that is cross-platform. And, to me, that platform includes Windows, not UWP.

**Comment by greatawesome** [comment]
> @MuleaneEve I mentioned I was initially skeptical about UNO Platform, not MAUI. MAUI I am pretty sure it's the most horrible announcement I've heard since I started using C# (around 12 years ago). Not because the one UI is a bad idea, it's because using Xamarin is a bad/horrible/creepy idea. (I'm lacking enough "bad" synonyms here and yeah, I'm pissed as many people are pissed with these)
> 
> Nowadays I think the UNO Platform guys are doing exactly what every C# developer expected in his whole life, XAML+C# which can be used on everything: Windows desktop (the only desktop OS people care about), Web, Android and iOS. Not this Xamarin/MAUI hybrid monster which aggregates no value.
> 
> And to be honest, in the years since I started programming, I've seen little improvement in the tools I use.
> Ok, .NET Core which runs on Linux ok, cool but except that, what? Oh, ASP.NET, why the hell I still need to use HTML/CSS instead of XAML? Is it something to punish the developers so they need to learn many coding standards? Is it too hard to use one thing everywhere? The UNO guys are doing that, so I would expect this as the minimum level from a corporation.
> 
> And I'm not asking MSFT to listen everything the public asks, as this is impossible because each one has one opinion. I'm asking to listen to the majority.

Uno is built on the work of the Xamarin team (just not the Xamarin.Forms portion of it). They used Xamarin to make WinUI/UWP cross platform. It is the Xamarin.Forms part of Xamarin that doesn't get as much love from some folks. That is partly why Maui isn't just a renaming of Xamarin.Forms, but also rewriting a lot of the back end code. Whether that'll please the critics remains to be seen.

An interesting development from Uno is a SkiaSharp backend, which they're using to run Uno apps on Linux. They already support running Xamarin.Forms apps that target UWP on their WASM back end, so I wonder if the same will be true for their Skia backend? It could be that Xamarin.Forms apps rendered in Skia is already very close, even if the route to get there is a bit convoluted.

**Comment by jcbeppler** [comment]
@greatawesome I think this image below shows why I like UNO Platform idea (not that I have used it so far). They use Xamarin strictly on IOs and Android, they don't mess up UWP.

UWP is excellent the way it is today, an environment which works and on which .NET developers are finally able to do native compilation for IP protection.

![image](https://user-images.githubusercontent.com/59994318/90308554-b3467080-dee0-11ea-8e3b-9a42bc96a406.png)
